### PR TITLE
gh-118830: Bump DEFAULT_PROTOCOL to 5 (#118830)

### PR DESCRIPTION
**Pull request by mdkcore0**
I didn't have add a news entry, as it is not a user facing change. Should do I add it?

* Issue: gh-118830

----
📚 Documentation preview 📚: https://cpython-previews--119340.org.readthedocs.build/

### CODE REVIEW

**Comment by mdkcore0** [comment]
On a second thought it makes sense to add it, but in which section this change falls on?

**Comment by carljm** [comment]
Yes, this deserves a NEWS entry, and it should go in `Library` section, I think.

**Comment by ambv** [comment]
This not only deserves a change log entry but also a "What's New in Python 3.14" entry.

**Comment by mdkcore0** [comment]
@carljm, @ambv: I've added requested entries

For the "WhatsNew in 3.14" I looked on how it was added for python 3.8 (the rst file), it includes some wording about the impact on performance (under Optimizations section):
> * The default protocol in the :mod:`pickle` module is now Protocol 4,
  first introduced in Python 3.4.  It offers better performance and smaller
  size compared to Protocol 3 available since Python 3.0.

Is it a good idea to add something like that here? Since I'm starting contributing now, I will need some help to fulfill it :)

File: Doc/library/pickle.rst [hunk 1]
```diff
@@ -199,7 +199,7 @@ The :mod:`pickle` module provides the following constants:
 
    An integer, the default :ref:`protocol version <pickle-protocols>` used
    for pickling.  May be less than :data:`HIGHEST_PROTOCOL`.  Currently the
-   default protocol is 4, first introduced in Python 3.4 and incompatible
+   default protocol is 5, first introduced in Python 3.4 and incompatible
```
Comment (carljm) [inline Doc/library/pickle.rst]: Did you check that protocol version 5 was also introduced in Python 3.4?

Finding where it was introduced might also uncover some information about how it is better than protocol 4, to add to What's New?

File: Doc/whatsnew/3.14.rst [hunk 2]
```diff
@@ -97,6 +97,7 @@ Added :func:`ast.compare` for comparing two ASTs.
 Optimizations
 =============
 
+* The default protocol in the :mod:`pickle` module is now Protocol 5.
```
Comment (carljm) [inline Doc/whatsnew/3.14.rst]: I don't personally have context to add here. Maybe another reviewer does.

File: Doc/library/pickle.rst (hunk 1 above)
Comment (mdkcore0) [reply Doc/library/pickle.rst]: > Did you check that protocol version 5 was also introduced in Python 3.4?
> 
> Finding where it was introduced might also uncover some information about how it is better than protocol 4, to add to What's New?

good point, found here it was added for python 3.8 (commit c879ff2); will pay attention to it to write a propert What's New, thanks! (also sorry for the delay, got some days off, but now I'm getting back to my backlog of contributions :))

File: Doc/whatsnew/3.14.rst (hunk 2 above)
Comment (picnixz) [reply Doc/whatsnew/3.14.rst]: Maybe "Set the default protocol version of the <code>:mod:\`pickle\`</code> module to ``5``"? Or mention the corresponding constant (if any)?

File: Modules/_pickle.c
Comment (devdanzin) [inline Modules/_pickle.c:1]: I believe you also need to change the default protocol in this file:

https://github.com/python/cpython/blob/ee0692a59121cdcd51533cec5716df308abcf8f4/Modules/_pickle.c#L41-L44

File: Modules/_pickle.c [hunk 3]
```diff
@@ -4693,7 +4693,7 @@ This takes a binary file for writing a pickle data stream.
 
 The optional *protocol* argument tells the pickler to use the given
 protocol; supported protocols are 0, 1, 2, 3, 4 and 5.  The default
-protocol is 4. It was introduced in Python 3.4, and is incompatible
+protocol is 5. It was introduced in Python 3.4, and is incompatible
```
Comment (devdanzin) [inline Modules/_pickle.c]: Same for all mentions of being introduced in 3.4:
```suggestion
protocol is 5. It was introduced in Python 3.8, and is incompatible
```

File: Modules/_pickle.c
Comment (mdkcore0) [reply Modules/_pickle.c:1]: didn't have find that one, thanks!

File: Doc/whatsnew/3.14.rst (hunk 2 above)
Comment (mdkcore0) [reply Doc/whatsnew/3.14.rst]: @picnixz I've changed it to your suggestion; also, there is no constant that can be referenced here.

I've added some wording on pickle.rst that I can replicate here too ("This version introduces support for out-of-band buffers, where :pep:`3118`-compatible data can be transmitted separately from the main pickle stream."), what do you think?

Comment (picnixz) [reply Doc/whatsnew/3.14.rst]: Mmh. I think it's perhaps good to simply add something like "See :pep:\`3118\` for details." or soemthing along the lines or anything related to what Protocol 5 brings (if it's a section in the pickle page, you can refer to it maybe?)

Comment (mdkcore0) [reply Doc/whatsnew/3.14.rst]: oh, cool, will look to something like that (or use your suggestion as fallback)

Comment (mdkcore0) [reply Doc/whatsnew/3.14.rst]: I've mentioned pickle's "data stream format" section, it seems to be a very good source of detailed information about it :)

File: Modules/_pickle.c (hunk 3 above)
Comment (chris-eibl) [reply Modules/_pickle.c]: I suggest changing in [line 159](https://github.com/python/cpython/blob/main/Doc/library/pickle.rst?plain=1#L159)

> It is the default protocol starting with Python 3.8.

to

> This was the default protocol in Python 3.8--3.14.

Likewise, adding to [line 161](https://github.com/python/cpython/blob/main/Doc/library/pickle.rst?plain=1#L161)
> It is the default protocol starting with Python 3.14.

**Comment by picnixz** [comment]
Before bumping it to version 5, I would like to address #120380 since it could introduce breakage. At least, there was something that was not correctly handled for pickle's buffers.

File: Modules/_pickle.c (hunk 3 above)
Comment (mdkcore0) [reply Modules/_pickle.c]: > I suggest changing in [line 159](https://github.com/python/cpython/blob/main/Doc/library/pickle.rst?plain=1#L159)
> 
> > It is the default protocol starting with Python 3.8.
> 
> to
> 
> > This was the default protocol in Python 3.8--3.14.
> 
> Likewise, adding to [line 161](https://github.com/python/cpython/blob/main/Doc/library/pickle.rst?plain=1#L161)
> 
> > It is the default protocol starting with Python 3.14.

makes senses, I've added the suggested changes!

**Comment by mdkcore0** [comment]
> Before bumping it to version 5, I would like to address #120380 since it could introduce breakage. At least, there was something that was not correctly handled for pickle's buffers.
no problem, thanks!!

**Comment by picnixz** [comment]
Now that https://github.com/python/cpython/issues/120380 is fixed, I think this can be done, unless someone finds something that is broken for protocol 5 or something that should be supported for protocol 5 and that is not. There are https://github.com/python/cpython/issues/84895 and https://github.com/python/cpython/issues/89467 which could be something that is desired before bumping the default protocol to protocol 5 since they could dramatically improve performances in general.

**Comment by encukou** [comment]
I don't think either of those needs to block this PR.

I've resolved the conflict in the What's New document, and in the process moved the entry from *Optimizations* to *Improved Modules*. Unlike version 4, version 5 isn't primarily a performance improvement.

**Review by encukou — APPROVED** [review]

File: Doc/library/pickle.rst [hunk 4]
```diff
@@ -156,13 +156,14 @@ to read the pickle produced.
 
 * Protocol version 4 was added in Python 3.4.  It adds support for very large
   objects, pickling more kinds of objects, and some data format
-  optimizations.  It is the default protocol starting with Python 3.8.
+  optimizations.  This was the default protocol in Python 3.8--3.14.
```
Comment (hugovk) [inline Doc/library/pickle.rst]:
```suggestion
  optimizations.  This was the default protocol in Python 3.8--3.13.
```

Comment (mdkcore0) [reply Doc/library/pickle.rst]: fixed!

**Comment by hugovk** [comment]
See also the [failed CI check](https://github.com/python/cpython/actions/runs/9779334010/job/26998341670?pr=119340):

> Generated files not up to date.
> Perhaps you forgot to run `make regen-all` or `build.bat --regen`. ;)

**Comment by mdkcore0** [comment]
> See also the [failed CI check](https://github.com/python/cpython/actions/runs/9779334010/job/26998341670?pr=119340):
> 
> > Generated files not up to date.
> > Perhaps you forgot to run `make regen-all` or `build.bat --regen`. ;)

oh, cool
I run the mentioned command, should I commit the modified files them (Modules/_pickle.c and Modules/clinic/_pickle.c.h)?

**Comment by hugovk** [comment]
Yes please!

**Comment by encukou** [comment]
Thank you!

**Pull request merged**